domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,118 @@
+        scorePerSkillMapping[linkedSkillId]?.let {
+          it.pointsReceived += questionScore
+          it.totalPointsAvailable += maxScorePerQuestion
+        }
+      }
+    }
+  }
+
+  /** Compute the overall score as well as the score and mastery per skill. */
+  internal fun computeAll(): UserAssessmentPerformance {
+    calculateScores()
+    // TODO(#3067): Create mastery calculations method
+
+    // Set up the return values
+    val finalScore = FractionGrade.newBuilder().apply {
+      pointsReceived = totalScore.pointsReceived.toDouble() / maxScorePerQuestion
+      totalPointsAvailable = totalScore.totalPointsAvailable.toDouble() / maxScorePerQuestion
+    }.build()
+
+    val finalScorePerSkillMapping = mutableMapOf<String, FractionGrade>()
+    for (score in scorePerSkillMapping) {
+      if (score.value.totalPointsAvailable != 0) { // Exclude entries with denominator of 0
+        finalScorePerSkillMapping[score.key] = FractionGrade.newBuilder().apply {
+          pointsReceived = score.value.pointsReceived.toDouble() / maxScorePerQuestion
+          totalPointsAvailable = score.value.totalPointsAvailable.toDouble() / maxScorePerQuestion
+        }.build()
+      }
+    }
+
+    // TODO(#3067): Set up finalMasteryPerSkillMapping
+    val finalMasteryPerSkillMapping = mapOf<String, Double>()
+
+    return UserAssessmentPerformance.newBuilder().apply {
+      totalFractionScore = finalScore
+      putAllFractionScorePerSkillMapping(finalScorePerSkillMapping)
+      putAllMasteryPerSkillMapping(finalMasteryPerSkillMapping)
+    }.build()
+  }
+
+  /**
+   * Mutable class that stores a grade as a fraction.
+   */
+  class MutableFractionGrade(
+    internal var pointsReceived: Int,
+    internal var totalPointsAvailable: Int
+  )
+
+  /**
+   * Factory to create a new [QuestionAssessmentCalculation].
+   */
+  class Factory @Inject constructor(
+    @ViewHintScorePenalty private val viewHintPenalty: Int,
+    @WrongAnswerScorePenalty private val wrongAnswerPenalty: Int,
+    @MaxScorePerQuestion private val maxScorePerQuestion: Int,
+  ) {
+    /**
+     * Creates a new [QuestionAssessmentCalculation] with its state set up.
+     *
+     * @param skillIdList the list of IDs for the skills that were tested in this practice session
+     * @param questionSessionMetrics metrics for the user's answers submitted, hints viewed, and
+     *        solutions viewed per question during this practice session
+     */
+    fun create(
+      skillIdList: List<String>,
+      questionSessionMetrics: List<QuestionSessionMetrics>
+    ): QuestionAssessmentCalculation {
+      // Set up the data structures needed for computing the user's scores
+      val totalScore = MutableFractionGrade(0, 0)
+      val scorePerSkillMapping = mutableMapOf<String, MutableFractionGrade>()
+      for (skillId in skillIdList) {
+        scorePerSkillMapping[skillId] = MutableFractionGrade(0, 0)
+      }
```

Comment:
Similarly, this can be done in a more Kotlin-idiomatic way (to avoid the mutable map):

```kotlin
skillIdList.associateWith { MutableFractionGrade() }
```

(I also suggest updating the constructor for MutableFractionGrade to default params to 0 to simplify things since it doesn't seem like we actually need to ever initialize with non-zero values).

Reply:
Fixed